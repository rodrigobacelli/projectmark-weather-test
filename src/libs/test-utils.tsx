import * as React from 'react';
import { render, RenderOptions } from '@testing-library/react';

const Providers = ({ children }: { children: React.ReactNode }) => {
  return children;
};

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: Providers, ...options });

export * from '@testing-library/react';
export { customRender as render };
