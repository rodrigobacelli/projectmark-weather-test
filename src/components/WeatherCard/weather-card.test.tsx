import { describe, it, expect } from 'vitest';
import { render } from '../../tests/test-utils';

import { WeatherCard } from './';
import { mockedWeather } from '../../tests/mocks/mockedWeather.ts';

describe('components/WeatherCard', () => {
  it('should render the city name', () => {
    const result = render(
      <WeatherCard weather={mockedWeather} className="weather-card-className" />
    );

    expect(result.getByText('San Francisco, US')).toBeInTheDocument();
  });

  it('should render the main weather condition', () => {
    const result = render(
      <WeatherCard weather={mockedWeather} className="weather-card-className" />
    );

    expect(result.getByText('Clouds')).toBeInTheDocument();
  });

  it('should render the temperature, humidity and pressure', () => {
    const result = render(
      <WeatherCard weather={mockedWeather} className="weather-card-className" />
    );

    expect(result.getByTestId('weather-card-humidity')).toHaveTextContent(
      '90%'
    );
    expect(result.getByTestId('weather-card-pressure')).toHaveTextContent(
      '1021'
    );
    expect(result.getByTestId('weather-card-temperature')).toHaveTextContent(
      '11'
    );
  });

  describe('temperture color rules', () => {
    it('should render the temperature as blue when 5 deg or less', () => {
      const result = render(
        <WeatherCard
          weather={{
            ...mockedWeather,
            main: {
              ...mockedWeather.main,
              temp: 4,
            },
          }}
          className="weather-card-className"
        />
      );

      expect(
        result
          .getByTestId('weather-card-temperature')
          .classList.contains('text-blue-900')
      ).toBeTruthy();
      expect(
        result
          .getByTestId('weather-card-temperature')
          .classList.contains('text-orange-900')
      ).toBeFalsy();
      expect(
        result
          .getByTestId('weather-card-temperature')
          .classList.contains('text-red-700')
      ).toBeFalsy();
    });

    it('should render the temperature as orange when above 5 deg, up to 25 deg', () => {
      const result = render(
        <WeatherCard
          weather={{
            ...mockedWeather,
            main: {
              ...mockedWeather.main,
              temp: 20,
            },
          }}
          className="weather-card-className"
        />
      );

      expect(
        result
          .getByTestId('weather-card-temperature')
          .classList.contains('text-blue-900')
      ).toBeFalsy();
      expect(
        result
          .getByTestId('weather-card-temperature')
          .classList.contains('text-orange-900')
      ).toBeTruthy();
      expect(
        result
          .getByTestId('weather-card-temperature')
          .classList.contains('text-red-700')
      ).toBeFalsy();
    });

    it('should render the temperature as red when above 25 deg', () => {
      const result = render(
        <WeatherCard
          weather={{
            ...mockedWeather,
            main: {
              ...mockedWeather.main,
              temp: 30,
            },
          }}
          className="weather-card-className"
        />
      );

      expect(
        result
          .getByTestId('weather-card-temperature')
          .classList.contains('text-blue-900')
      ).toBeFalsy();
      expect(
        result
          .getByTestId('weather-card-temperature')
          .classList.contains('text-orange-900')
      ).toBeFalsy();
      expect(
        result
          .getByTestId('weather-card-temperature')
          .classList.contains('text-red-700')
      ).toBeTruthy();
    });
  });

  it('should accept classNames', () => {
    const result = render(
      <WeatherCard weather={mockedWeather} className="weather-card-className" />
    );

    expect(
      result.container.querySelector('.weather-card-className')
    ).toBeInTheDocument();
  });

  it('should render nothing when weather is empty', () => {
    const result = render(<WeatherCard className="weather-card-className" />);

    expect(!!result.queryByTestId('weather-card-temperature')).toBeFalsy();
  });

  it('should render the Weather Card Component', () => {
    const result = render(
      <WeatherCard weather={mockedWeather} className="weather-card-className" />
    );

    expect(result).toMatchSnapshot();
  });
});
