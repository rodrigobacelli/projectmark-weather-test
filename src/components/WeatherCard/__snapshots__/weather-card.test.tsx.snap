// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/WeatherCard > should render the Weather Card Component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="p-4 bg-gray-100 dark:bg-gray-800 rounded-lg ring shadow-xl ring-gray-900/5 w-full flex flex-col gap-4 weather-card-className"
        data-testid="weather-card"
      >
        <h2
          class="text-gray-900 dark:text-gray-100 flex items-center gap-1 text-sm"
        >
          <svg
            class="w-3 h-3"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 384 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Location
            </title>
            <path
              d="M215.7 499.2C267 435 384 279.4 384 192C384 86 298 0 192 0S0 86 0 192c0 87.4 117 243 168.3 307.2c12.3 15.3 35.1 15.3 47.4 0zM192 128a64 64 0 1 1 0 128 64 64 0 1 1 0-128z"
            />
          </svg>
          San Francisco, US
        </h2>
        <article
          class="flex items-center justify-between gap-1"
        >
          <p
            class="text-7xl flex justify-start gap-1 text-orange-900 dark:text-orange-600"
            data-testid="weather-card-temperature"
          >
            11
            <svg
              class="w-8 h-8 mt-2"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4.5 10C2.567 10 1 8.433 1 6.5C1 4.567 2.567 3 4.5 3C6.433 3 8 4.567 8 6.5C8 8.433 6.433 10 4.5 10ZM4.5 8C5.32843 8 6 7.32843 6 6.5C6 5.67157 5.32843 5 4.5 5C3.67157 5 3 5.67157 3 6.5C3 7.32843 3.67157 8 4.5 8ZM22 10H20C20 7.79086 18.2091 6 16 6C13.7909 6 12 7.79086 12 10V15C12 17.2091 13.7909 19 16 19C18.2091 19 20 17.2091 20 15H22C22 18.3137 19.3137 21 16 21C12.6863 21 10 18.3137 10 15V10C10 6.68629 12.6863 4 16 4C19.3137 4 22 6.68629 22 10Z"
              />
            </svg>
          </p>
          <div
            class="text-gray-900 dark:text-gray-100 flex flex-col gap-1"
          >
            <svg
              class="h-10 w-10"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 640 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M294.2 1.2c5.1 2.1 8.7 6.7 9.6 12.1l14.1 84.7 84.7 14.1c5.4 .9 10 4.5 12.1 9.6s1.5 10.9-1.6 15.4l-38.5 55c-2.2-.1-4.4-.2-6.7-.2c-23.3 0-45.1 6.2-64 17.1l0-1.1c0-53-43-96-96-96s-96 43-96 96s43 96 96 96c8.1 0 15.9-1 23.4-2.9c-36.6 18.1-63.3 53.1-69.8 94.9l-24.4 17c-4.5 3.2-10.3 3.8-15.4 1.6s-8.7-6.7-9.6-12.1L98.1 317.9 13.4 303.8c-5.4-.9-10-4.5-12.1-9.6s-1.5-10.9 1.6-15.4L52.5 208 2.9 137.2c-3.2-4.5-3.8-10.3-1.6-15.4s6.7-8.7 12.1-9.6L98.1 98.1l14.1-84.7c.9-5.4 4.5-10 9.6-12.1s10.9-1.5 15.4 1.6L208 52.5 278.8 2.9c4.5-3.2 10.3-3.8 15.4-1.6zM144 208a64 64 0 1 1 128 0 64 64 0 1 1 -128 0zM639.9 431.9c0 44.2-35.8 80-80 80l-271.9 0c-53 0-96-43-96-96c0-47.6 34.6-87 80-94.6l0-1.3c0-53 43-96 96-96c34.9 0 65.4 18.6 82.2 46.4c13-9.1 28.8-14.4 45.8-14.4c44.2 0 80 35.8 80 80c0 5.9-.6 11.7-1.9 17.2c37.4 6.7 65.8 39.4 65.8 78.7z"
              />
            </svg>
             
            Clouds
          </div>
        </article>
        <footer
          class="flex flex gap-4 items-center justify-start"
        >
          <p
            class="text-gray-900 dark:text-gray-100 flex gap-2 items-center"
            data-testid="weather-card-humidity"
          >
            <svg
              class="w-4 h-4"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 384 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title>
                Humidity
              </title>
              <path
                d="M192 512C86 512 0 426 0 320C0 228.8 130.2 57.7 166.6 11.7C172.6 4.2 181.5 0 191.1 0l1.8 0c9.6 0 18.5 4.2 24.5 11.7C253.8 57.7 384 228.8 384 320c0 106-86 192-192 192zM96 336c0-8.8-7.2-16-16-16s-16 7.2-16 16c0 61.9 50.1 112 112 112c8.8 0 16-7.2 16-16s-7.2-16-16-16c-44.2 0-80-35.8-80-80z"
              />
            </svg>
            90
            %
          </p>
          <div
            class="flex gap-2 items-center"
          >
            <svg
              class="w-4 h-4 text-gray-900 dark:text-gray-100"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <title>
                Atmospheric Pressure
              </title>
              <path
                d="M12 2C17.5228 2 22 6.47715 22 12C22 17.5228 17.5228 22 12 22C6.47715 22 2 17.5228 2 12C2 6.47715 6.47715 2 12 2ZM12 4C7.58172 4 4 7.58172 4 12C4 16.4183 7.58172 20 12 20C16.4183 20 20 16.4183 20 12C20 7.58172 16.4183 4 12 4ZM15.8329 7.33748C16.0697 7.17128 16.3916 7.19926 16.5962 7.40381C16.8002 7.60784 16.8267 7.92955 16.6587 8.16418C14.479 11.2095 13.2796 12.8417 13.0607 13.0607C12.4749 13.6464 11.5251 13.6464 10.9393 13.0607C10.3536 12.4749 10.3536 11.5251 10.9393 10.9393C11.3126 10.5661 12.9438 9.36549 15.8329 7.33748ZM17.5 11C18.0523 11 18.5 11.4477 18.5 12C18.5 12.5523 18.0523 13 17.5 13C16.9477 13 16.5 12.5523 16.5 12C16.5 11.4477 16.9477 11 17.5 11ZM6.5 11C7.05228 11 7.5 11.4477 7.5 12C7.5 12.5523 7.05228 13 6.5 13C5.94772 13 5.5 12.5523 5.5 12C5.5 11.4477 5.94772 11 6.5 11ZM8.81802 7.40381C9.20854 7.79433 9.20854 8.4275 8.81802 8.81802C8.4275 9.20854 7.79433 9.20854 7.40381 8.81802C7.01328 8.4275 7.01328 7.79433 7.40381 7.40381C7.79433 7.01328 8.4275 7.01328 8.81802 7.40381ZM12 5.5C12.5523 5.5 13 5.94772 13 6.5C13 7.05228 12.5523 7.5 12 7.5C11.4477 7.5 11 7.05228 11 6.5C11 5.94772 11.4477 5.5 12 5.5Z"
              />
            </svg>
            <p
              class="text-gray-900 dark:text-gray-100"
              data-testid="weather-card-pressure"
            >
              1021
               
              <abbr
                title="hectoPascals"
              >
                hPa
              </abbr>
            </p>
          </div>
        </footer>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="p-4 bg-gray-100 dark:bg-gray-800 rounded-lg ring shadow-xl ring-gray-900/5 w-full flex flex-col gap-4 weather-card-className"
      data-testid="weather-card"
    >
      <h2
        class="text-gray-900 dark:text-gray-100 flex items-center gap-1 text-sm"
      >
        <svg
          class="w-3 h-3"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 384 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            Location
          </title>
          <path
            d="M215.7 499.2C267 435 384 279.4 384 192C384 86 298 0 192 0S0 86 0 192c0 87.4 117 243 168.3 307.2c12.3 15.3 35.1 15.3 47.4 0zM192 128a64 64 0 1 1 0 128 64 64 0 1 1 0-128z"
          />
        </svg>
        San Francisco, US
      </h2>
      <article
        class="flex items-center justify-between gap-1"
      >
        <p
          class="text-7xl flex justify-start gap-1 text-orange-900 dark:text-orange-600"
          data-testid="weather-card-temperature"
        >
          11
          <svg
            class="w-8 h-8 mt-2"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4.5 10C2.567 10 1 8.433 1 6.5C1 4.567 2.567 3 4.5 3C6.433 3 8 4.567 8 6.5C8 8.433 6.433 10 4.5 10ZM4.5 8C5.32843 8 6 7.32843 6 6.5C6 5.67157 5.32843 5 4.5 5C3.67157 5 3 5.67157 3 6.5C3 7.32843 3.67157 8 4.5 8ZM22 10H20C20 7.79086 18.2091 6 16 6C13.7909 6 12 7.79086 12 10V15C12 17.2091 13.7909 19 16 19C18.2091 19 20 17.2091 20 15H22C22 18.3137 19.3137 21 16 21C12.6863 21 10 18.3137 10 15V10C10 6.68629 12.6863 4 16 4C19.3137 4 22 6.68629 22 10Z"
            />
          </svg>
        </p>
        <div
          class="text-gray-900 dark:text-gray-100 flex flex-col gap-1"
        >
          <svg
            class="h-10 w-10"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 640 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M294.2 1.2c5.1 2.1 8.7 6.7 9.6 12.1l14.1 84.7 84.7 14.1c5.4 .9 10 4.5 12.1 9.6s1.5 10.9-1.6 15.4l-38.5 55c-2.2-.1-4.4-.2-6.7-.2c-23.3 0-45.1 6.2-64 17.1l0-1.1c0-53-43-96-96-96s-96 43-96 96s43 96 96 96c8.1 0 15.9-1 23.4-2.9c-36.6 18.1-63.3 53.1-69.8 94.9l-24.4 17c-4.5 3.2-10.3 3.8-15.4 1.6s-8.7-6.7-9.6-12.1L98.1 317.9 13.4 303.8c-5.4-.9-10-4.5-12.1-9.6s-1.5-10.9 1.6-15.4L52.5 208 2.9 137.2c-3.2-4.5-3.8-10.3-1.6-15.4s6.7-8.7 12.1-9.6L98.1 98.1l14.1-84.7c.9-5.4 4.5-10 9.6-12.1s10.9-1.5 15.4 1.6L208 52.5 278.8 2.9c4.5-3.2 10.3-3.8 15.4-1.6zM144 208a64 64 0 1 1 128 0 64 64 0 1 1 -128 0zM639.9 431.9c0 44.2-35.8 80-80 80l-271.9 0c-53 0-96-43-96-96c0-47.6 34.6-87 80-94.6l0-1.3c0-53 43-96 96-96c34.9 0 65.4 18.6 82.2 46.4c13-9.1 28.8-14.4 45.8-14.4c44.2 0 80 35.8 80 80c0 5.9-.6 11.7-1.9 17.2c37.4 6.7 65.8 39.4 65.8 78.7z"
            />
          </svg>
           
          Clouds
        </div>
      </article>
      <footer
        class="flex flex gap-4 items-center justify-start"
      >
        <p
          class="text-gray-900 dark:text-gray-100 flex gap-2 items-center"
          data-testid="weather-card-humidity"
        >
          <svg
            class="w-4 h-4"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 384 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Humidity
            </title>
            <path
              d="M192 512C86 512 0 426 0 320C0 228.8 130.2 57.7 166.6 11.7C172.6 4.2 181.5 0 191.1 0l1.8 0c9.6 0 18.5 4.2 24.5 11.7C253.8 57.7 384 228.8 384 320c0 106-86 192-192 192zM96 336c0-8.8-7.2-16-16-16s-16 7.2-16 16c0 61.9 50.1 112 112 112c8.8 0 16-7.2 16-16s-7.2-16-16-16c-44.2 0-80-35.8-80-80z"
            />
          </svg>
          90
          %
        </p>
        <div
          class="flex gap-2 items-center"
        >
          <svg
            class="w-4 h-4 text-gray-900 dark:text-gray-100"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Atmospheric Pressure
            </title>
            <path
              d="M12 2C17.5228 2 22 6.47715 22 12C22 17.5228 17.5228 22 12 22C6.47715 22 2 17.5228 2 12C2 6.47715 6.47715 2 12 2ZM12 4C7.58172 4 4 7.58172 4 12C4 16.4183 7.58172 20 12 20C16.4183 20 20 16.4183 20 12C20 7.58172 16.4183 4 12 4ZM15.8329 7.33748C16.0697 7.17128 16.3916 7.19926 16.5962 7.40381C16.8002 7.60784 16.8267 7.92955 16.6587 8.16418C14.479 11.2095 13.2796 12.8417 13.0607 13.0607C12.4749 13.6464 11.5251 13.6464 10.9393 13.0607C10.3536 12.4749 10.3536 11.5251 10.9393 10.9393C11.3126 10.5661 12.9438 9.36549 15.8329 7.33748ZM17.5 11C18.0523 11 18.5 11.4477 18.5 12C18.5 12.5523 18.0523 13 17.5 13C16.9477 13 16.5 12.5523 16.5 12C16.5 11.4477 16.9477 11 17.5 11ZM6.5 11C7.05228 11 7.5 11.4477 7.5 12C7.5 12.5523 7.05228 13 6.5 13C5.94772 13 5.5 12.5523 5.5 12C5.5 11.4477 5.94772 11 6.5 11ZM8.81802 7.40381C9.20854 7.79433 9.20854 8.4275 8.81802 8.81802C8.4275 9.20854 7.79433 9.20854 7.40381 8.81802C7.01328 8.4275 7.01328 7.79433 7.40381 7.40381C7.79433 7.01328 8.4275 7.01328 8.81802 7.40381ZM12 5.5C12.5523 5.5 13 5.94772 13 6.5C13 7.05228 12.5523 7.5 12 7.5C11.4477 7.5 11 7.05228 11 6.5C11 5.94772 11.4477 5.5 12 5.5Z"
            />
          </svg>
          <p
            class="text-gray-900 dark:text-gray-100"
            data-testid="weather-card-pressure"
          >
            1021
             
            <abbr
              title="hectoPascals"
            >
              hPa
            </abbr>
          </p>
        </div>
      </footer>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
